module/mempool/stdmap/backend.go

```diff
@@ -211,12 +226,17 @@ func (b *Backend) reduce() {
 		key, _ := b.eject(b.entities)
 
 		// if the key is not actually part of the map, use stupid fallback eject
-		_, ok := b.entities[key]
+		entity, ok := b.entities[key]
 		if !ok {
 			key, _ = EjectFakeRandom(b.entities)
 		}
 
 		// remove the key
 		delete(b.entities, key)
+
+		// notify callback
+		if b.ejectionCallback != nil {
```

Comment:
This is an optional suggestion. 

I was thinking to have a default `ejectionCallback` which basically does a noop, so that it will never be nil and we don't need to remember handle the nil case here. And we could also get rid of the RegisterEjectionCallback by taking the callback in the constructor argument:

```
func NewBackend(options ...OptionFunc) *Backend {
	b := Backend{
		Backdata:         NewBackdata(),
		limit:            uint(math.MaxUint32),
		eject:            EjectTrueRandom,
		ejectionCallback: func(entity *Entity) {}, // noop
	}
	for _, option := range options {
		option(&b)
        }
}
```

This would be consistent with how to provide default `limit` and default `eject` strategy, where we provide a default ejection callback.

So if we want to register a callback, we could do this instead:

```
	ensureEntityNotInMempool := func(entity flow.Entity) {
		id := entity.ID()
		go func() {
			e, found := pool.ByID(id)
			require.False(t, found)
			require.Nil(t, e)
		}()
		go func() {
			require.False(t, pool.Has(id))
		}()
	}

        pool := NewBackend(WithLimit(limit), withEjectionCallback(ensureEntityNotInMempool))
```

Reply:
Yea, I have been thinking about this pattern too. Lets take a look at a usage pattern, which I assume will be quite typical:
https://github.com/onflow/flow-go/blob/f1b96f4ed93d7da1c28a08bb1b943737e683a272/module/mempool/stdmap/incorporated_results.go#L22-L39

How would this look like, if we wanted to add the `EjectionCallback` at construction time to the backend? I could imagine writing something like: 
```golang
	mempool := &IncorporatedResults{
		size:    0,
	}

	adjustSizeOnEjection := func(entity flow.Entity) {
		// uncaught type assertion; should never panic as the mempool only stores IncorporatedResultMap:
		incorporatedResultMap := entity.(*model.IncorporatedResultMap)
		mempool.size -= uint(len(incorporatedResultMap.IncorporatedResults))
	}
	mempool.backend := NewBackend(WithLimit(limit), WithEjectionCallback(adjustSizeOnEjection))

	return mempool, nil
```

I feel this pattern also has a disadvantage
* I first need to construct a disfunctional struct: `mempool` without a Backend
* only thereafter, I can construct the backend, because the backend needs the ejector, which in turn needs the higher-level mempool 

Kind of a circular dependency. Allowing to register the ejection callback _after_ construction allows to break this circular object dependency. It is kind of inspired by the actor model. 

Either way. I think both solutions have advantages and drawbacks. Happy to implement your suggestion. Just wanted to share my thinking. Please let me know.

Reply:
After all, we specify the configurable parameters of the Backend through options. Therefore, I feel your suggestion increases consistency.

Reply:
Yeah, I see your point. 

You are right, a "disfunctional" mempool is constructed, but at least, it is only being "disfunctional" within the `NewIncorporatedResults`. No one else will have access to the "disfunctional" mempool object. And after `NewIncorporatedResults` is done, it will return a functional mempool.

So I think should be OK.